### PR TITLE
fix(sql): use correct message field as JSON request body

### LINKED ISSUE #12112: Cloud SQL Admin API: `CreateEphemeral` sends malformed API request
**Does this issue affect the google-cloud-cpp project?**
Yes

**What component of `google-cloud-cpp` is this related to?**
`google/cloud/sql`

**Describe the bug**
The `google::cloud::sql::SqlInstancesServiceClient::CreateEphemeral()` method doesn't work as expected - complaining about missing `public_key` when one is provided.

It seems that the request being made to the REST endpoint embeds the public key in `body` in JSON while the endpoint expects it to be a top-level field (see logs at the end). Apart from that, `ProtoRequestToJsonPayload` converts `public_key` to `publicKey` due to `preserve_proto_field_names` being `false` by default.

**To Reproduce**

```cpp
namespace sql = ::google::cloud::sql_v1;
auto client = sql::SqlInstancesServiceClient(
    google::cloud::ExperimentalTag{}, sql::MakeSqlInstancesServiceConnectionRest(google::cloud::ExperimentalTag{}));

google::cloud::sql::v1::SqlInstancesCreateEphemeralCertRequest create_req;
create_req.set_project("xxx");
create_req.set_instance("xxx");
create_req.mutable_body()->set_public_key("THE_PUBLIC_KEY");

auto eph_res = client.CreateEphemeral(create_req);

if (eph_res) {
  std::cout << eph_res->DebugString();
} else {
  std::cout << "Error: " << eph_res.status() << "\n";
}
```

**Expected behavior**
An ephemeral certificate is returned (as in [the api definition](https://cloud.google.com/sql/docs/postgres/admin-api/rest/v1/connect/generateEphemeralCert))

**Operating system:**
Ubuntu 22.04.2 LTS

**What compiler and version are you using?**
clang 16.0.5

**What version of `google-cloud-cpp` are you using?**
v2.13.0

**Additional context**
Selected curl logs from running with `GOOGLE_CLOUD_CPP_ENABLE_CLOG=DEBUG GOOGLE_CLOUD_CPP_ENABLE_TRACING=http`:

```log
(...)
>> curl(Send Header): POST /v1/projects/{project}/instances/{instance}/createEphemeral HTTP/2
Host: sqladmin.googleapis.com
User-Agent: gcloud-cpp/v2.13.0 (Clang-16.0.5; ex) gcloud-cpp/v2.13.0 (Clang-16.0.5; ex) libcurl/8.1.2-DEV OpenSSL/3.0.7 zlib/1.2.13 nghttp2/1.53.0
(...)
>> curl(Send Data): size=530 data={"instance":"xxx","project":"xxx","body":{"publicKey":"-----BEGIN PUBLIC KEY-----\nMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AM
== curl(Info): We are completely uploaded and fine
(...)
<< curl(Recv Data): size=235 data={.  "error": {.    "code": 400,.    "message": "Missing parameter: public_key.",.    "errors": [.      {.        "message": "Mis
(...)
Error: INVALID_ARGUMENT: Error in non-idempotent operation CreateEphemeral: Missing parameter: public_key. error_info={reason=required, domain=global, metadata={http_status_code=400}}
```

**Workaround**
The following dirty workaround rewrites the request and leads to a successful API response:

```patch
diff --git a/google/cloud/internal/rest_stub_helpers.cc b/google/cloud/internal/rest_stub_helpers.cc
index 272ed478..fad85e81 100644
--- a/google/cloud/internal/rest_stub_helpers.cc
+++ b/google/cloud/internal/rest_stub_helpers.cc
@@ -47,8 +47,10 @@ Status RestResponseToProto(google::protobuf::Message& destination,
 
 Status ProtoRequestToJsonPayload(google::protobuf::Message const& request,
                                  std::string& json_payload) {
-  auto proto_to_json_status =
-      google::protobuf::util::MessageToJsonString(request, &json_payload);
+  google::protobuf::util::JsonPrintOptions print_options;
+  print_options.preserve_proto_field_names = true;
+  auto proto_to_json_status = google::protobuf::util::MessageToJsonString(
+      request, &json_payload, print_options);
   if (!proto_to_json_status.ok()) {
     return internal::InternalError(
         std::string(proto_to_json_status.message()),
diff --git a/google/cloud/internal/rest_stub_helpers.h b/google/cloud/internal/rest_stub_helpers.h
index 14dc304a..638e5667 100644
--- a/google/cloud/internal/rest_stub_helpers.h
+++ b/google/cloud/internal/rest_stub_helpers.h
@@ -105,6 +105,13 @@ StatusOr<Response> Post(
   std::string json_payload;
   auto status = ProtoRequestToJsonPayload(request, json_payload);
   if (!status.ok()) return status;
+  if (path == "/v1/projects/{project}/instances/{instance}/createEphemeral") {
+    size_t start_position = json_payload.find("\"public_key\":\"") + 14;
+    std::string public_key = json_payload.substr(
+        start_position,
+        json_payload.find('"', start_position) - start_position);
+    json_payload = R"({"public_key":")" + public_key + "\"}";
+  }
   rest_internal::RestRequest rest_request;
   rest_request.SetPath(std::move(path));
   for (auto& p : query_params) {
diff --git a/google/cloud/sql/v1/internal/sql_instances_rest_stub.cc b/google/cloud/sql/v1/internal/sql_instances_rest_stub.cc
index 750bc5b3..6d04ec5e 100644
--- a/google/cloud/sql/v1/internal/sql_instances_rest_stub.cc
+++ b/google/cloud/sql/v1/internal/sql_instances_rest_stub.cc
@@ -242,7 +242,9 @@ DefaultSqlInstancesServiceRestStub::CreateEphemeral(
         request) {
   return rest_internal::Post<google::cloud::sql::v1::SslCert>(
       *service_, rest_context, request,
-      "/v1/projects/{project}/instances/{instance}/createEphemeral");
+      "/v1/projects/{project}/instances/{instance}/createEphemeral",
+      {std::make_pair("instance", request.instance()),
+       std::make_pair("project", request.project())});
 }
 
 StatusOr<google::cloud::sql::v1::Operation>

```

### PR DESCRIPTION
**Pull request by scotthart**
fixes #12112

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12207)

### CODE REVIEW

File: google/cloud/sql/integration_tests/sql_integration_test.cc [hunk 1]
```diff
@@ -54,6 +56,63 @@ TEST_F(SqlIntegrationTest, PaginatedList) {
...
+
+  google::cloud::sql::v1::SqlInstancesCreateEphemeralCertRequest request;
+  request.set_project(project_id_);
+  request.set_instance((*instance.begin())->name());
+  request.mutable_body()->set_public_key("THE_PUBLIC_KEY");
+
+  auto result = client.CreateEphemeral(request);
+  EXPECT_THAT(
+      result,
+      ::testing::AnyOf(
```
Comment (dbolduc) [inline google/cloud/sql/integration_tests/sql_integration_test.cc:106]: nit: `using ::testing::AnyOf;`

File: google/cloud/sql/integration_tests/sql_integration_test.cc [hunk 2]
```diff
@@ -54,6 +56,63 @@ TEST_F(SqlIntegrationTest, PaginatedList) {
...
+          StatusIs(StatusCode::kInvalidArgument,
+                   HasSubstr("Provided public key was in an invalid or "
+                             "unsupported format")),
+          (StatusIs(
```
Comment (dbolduc) [inline google/cloud/sql/integration_tests/sql_integration_test.cc:110]: nit: outer parentheses are unnecessary

File: google/cloud/sql/integration_tests/sql_integration_test.cc [hunk 3]
```diff
@@ -54,6 +56,63 @@ TEST_F(SqlIntegrationTest, PaginatedList) {
...
+  namespace sql = ::google::cloud::sql_v1;
+  auto client =
+      sql::SqlInstancesServiceClient(google::cloud::ExperimentalTag{},
+                                     sql::MakeSqlInstancesServiceConnectionRest(
+                                         google::cloud::ExperimentalTag{}));
+
+  google::cloud::sql::v1::SqlInstancesListRequest list_request;
+  list_request.set_project(project_id_);
+  auto instance = client.List(list_request);
+  ASSERT_STATUS_OK(*instance.begin());
```
Comment (dbolduc) [inline google/cloud/sql/integration_tests/sql_integration_test.cc:96]: I think you need to

```cc
ASSERT_NE(instance.begin(), instance.end());
```

... before accessing the value of the iterator.

and you probably want to s/instance/instances/

**Review by dbolduc — APPROVED** [review]

File: google/cloud/sql/integration_tests/sql_integration_test.cc (hunk 3 above)
Comment (coryan) [reply google/cloud/sql/integration_tests/sql_integration_test.cc:96]: Or `ASSERT_THAT(instance, Not(IsEmpty()))` or `EXPECT_THAT(instance, Contains(IsOkay()))` ?

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12207)* status: 0 of 13 files reviewed, 2 unresolved discussions (waiting on @coryan and @dbolduc)

___
*[`google/cloud/sql/integration_tests/sql_integration_test.cc` line 96 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12207#gh-1273066317:-NaCezPI48GoXZHLwcfh:b-fafklb) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/805f8e0e0ccac1217822f728899743437093b410/google/cloud/sql/integration_tests/sql_integration_test.cc#L96)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Or `ASSERT_THAT(instance, Not(IsEmpty()))` or `EXPECT_THAT(instance, Contains(IsOkay()))` ?
</blockquote></details>

g::c::StreamRange provides neither `empty()` nor `size()`, so the `IsEmpty` matcher doesn't work.
___
*[`google/cloud/sql/integration_tests/sql_integration_test.cc` line 106 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12207#gh-1273050734:-NaCfCDK90r7zmWhUK8s:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/805f8e0e0ccac1217822f728899743437093b410/google/cloud/sql/integration_tests/sql_integration_test.cc#L106)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

nit: `using ::testing::AnyOf;`
</blockquote></details>

Done.
___
*[`google/cloud/sql/integration_tests/sql_integration_test.cc` line 110 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12207#gh-1273052099:-NaCfCwa3nYhCBNEkWvp:bss4a2b) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/805f8e0e0ccac1217822f728899743437093b410/google/cloud/sql/integration_tests/sql_integration_test.cc#L110)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

nit: outer parentheses are unnecessary
</blockquote></details>

Removed.

**Pull request merged**